File: src/Suduko.java
//
public class Suduko {
	private int[][] suduko;
	
	public Suduko (int[][] suduko) {
		this.suduko = suduko;
	}
	
	public boolean recursionSearch () {
		int rowIndex = startPoint().rowIndex;
		int columnIndex = startPoint().columnIndex;
		
		if (rowIndex >= 8 && columnIndex >= 8 && suduko[rowIndex][columnIndex] != 0) {
			return true;
		}
		
		for (int value = 1; value < 10; value++) {
			if (checkAll(rowIndex, columnIndex, value)) {
				suduko[rowIndex][columnIndex] = value;
				if (!recursionSearch()) {
					suduko[rowIndex][columnIndex] = 0;
				}
				else {
					return true;
				}
			}
		}
		
		return false;
	}

//	public void print () {
//		for (int i = 0; i < 9; i++) {
//			for (int j = 0; j < 9; j++) {
//				System.out.print(suduko[i][j] + " ");
//			}
//			System.out.println();
//		}
//	}

	private boolean checkRow (int rowIndex, int value) {
		for (int i = 0; i < 9; i++) {
			if (suduko[rowIndex][i] == value) {
				return false;
			}
		}
		return true;
	}
	
	private boolean checkColumn (int columnIndex, int value) {
		for (int i = 0; i < 9; i++) {
			if (suduko[i][columnIndex] == value) {
				return false;
			}
		}
		return true;
	}
	
	private boolean checkSquare (int rowIndex, int columnIndex, int value) {
		for (int i = 3 * (rowIndex / 3); i < 3 * (rowIndex / 3) + 3; i++) {
			for (int j = 3 * (columnIndex / 3); j < 3 * (columnIndex / 3) + 3; j++) {
				if (suduko[i][j] == value) {
					return false;
				}
			}
		}
		return true;
	}
	
	private boolean checkAll (int rowIndex, int columnIndex, int value) {
		return checkRow(rowIndex, value) && checkColumn(columnIndex, value) && checkSquare(rowIndex, columnIndex, value);
	}
	
	private IndexPair startPoint() {
		int i = 0, j = 0;
		IndexPair indexPair = new IndexPair();
		
		for (i = 0; i < 9; i++) {
			for (j = 0; j < 9; j++) {
				if (suduko[i][j] == 0) {
					indexPair.rowIndex = i;
					indexPair.columnIndex = j;
					return indexPair;
				}
			}
		}
		
		indexPair.rowIndex = i - 1;
		indexPair.columnIndex = j - 1;
		return indexPair;
	}
	
	private class IndexPair {
		public int rowIndex, columnIndex;
	}
}

File: Suduko.py
class Suduko(object):
    
    def __init__(self, s):
        self.s = s
    
    def check_all(self, i, j, value):
        return self.check_row(i, j, value) and self.check_column(i, j, value) and self.check_small_sudoku(i, j, value)
    
    def check_row(self, i, j, value):
        return value not in self.s[i]
    
    def check_column(self, i, j, value):
        column = [self.s[v][j] for v in range(9)]
        return value not in column
    
    def check_small_sudoku(self, i, j, value):
        small_sudoku = [self.s[m][n] for m in range(i/3*3,(i/3+1)*3) for n in range(j/3*3,(j/3+1)*3)]
        return value not in small_sudoku
    
    def recursion_search(self):
        i,j = self.start_point()
        if i >=8 and j >=8 and self.s[8][8]:
            return True
        
        for value in range(1,10):
            if self.check_all(i, j, value):
                self.s[i][j] = value
                if not self.recursion_search():
                    self.s[i][j] = 0
                else:
                    return True
        return False 
    
    def start_point(self):
        for i in range(9):
            for j in range(9):
                if not self.s[i][j]:
                    return i,j
        return i,j

if '__main__' == __name__:
    
    s = [[6,0,3,7,0,0,0,0,0],
         [0,1,0,0,3,0,0,0,4],
         [0,8,0,9,5,6,0,0,3],
         [4,0,2,3,8,0,0,0,1],
         [1,0,0,2,0,5,0,0,0],
         [0,0,7,4,6,0,0,5,0],
         [2,9,1,0,0,0,7,8,0],
         [8,4,0,0,0,0,2,0,6],
         [0,7,0,5,2,0,0,1,9]]

    S = Suduko(s)
         
    S.recursion_search()
         
    for i in range(9):
        print S.s[i]